Dynamic datasources support

vadim-gleyf opened the pull request:
## what
* Now use url from arg `datasource` if it passed
* Regenerate readme

## why
* https://github.com/cloudposse/build-harness/issues/87

### Depend on [#55](https://github.com/cloudposse/packages/pull/55)
* Need to bump `packages` [version](https://github.com/cloudposse/build-harness/blob/master/modules/packages/Makefile#L2) when it merged 

resolve: #87

vadim-gleyf committed:
feat(modules/readme): dynamic datasources support

Now use url from arg datasource if it passed
Regenerate readme

osterman reviewed (changes requested):
The objective is to move all data source definitions into the readme template. That way, we can simply pass —file and —out and it works. No obligation on the caller to provide data sources.

vadim-gleyf committed:
Merge branch 'master' into dynamic-datasources

vadim-gleyf committed:
chore(packages/Makefile): bump packages version

vadim-gleyf committed:
feat(readme): add inline datasources support

Now datasource for readme defines in template
Also exist possibility override datasource due env variable

vadim-gleyf committed:
Merge branch 'master' into dynamic-datasources

osterman commented on templates/README.md:
```diff
@@ -1,5 +1,8 @@
 <!-- This file was automatically generated by the `build-harness`. Make all changes to `README.yaml` and run `make readme` to rebuild this file. -->
-
+<!-- 
+{{ defineDatasource "config" .Env.DATASOURCE }}
```
Use {{- ... -}} instead of the HTML comments

osterman commented on modules/readme/Makefile:
```diff
@@ -3,6 +3,9 @@ export README_FILE ?= README.md
 export README_YAML ?= README.yaml
 export README_TEMPLATE_FILE ?= $(BUILD_HARNESS_PATH)/templates/README.md
 export README_TEMPLATE_YAML ?= $(BUILD_HARNESS_PATH)/templates/README.yaml
+export USED_DATASOURCE ?= $(if $(datasource),$(datasource),$(shell pwd)/$(README_YAML))
```
What is this for?

vadim-gleyf replied:
@osterman  For correct output 
`@echo "Generated $(README_FILE) from $(README_TEMPLATE_FILE) using data from $(USED_DATASOURCE)"`

vadim-gleyf committed:
Merge branch 'master' into dynamic-datasources

vadim-gleyf committed:
fix(Makefile): clean up output of defineDatasource

vadim-gleyf committed:
Merge branch 'dynamic-datasources' of github.com:cloudposse/build-harness into dynamic-datasources

osterman commented on templates/README.md:
```diff
@@ -1,5 +1,6 @@
 <!-- This file was automatically generated by the `build-harness`. Make all changes to `README.yaml` and run `make readme` to rebuild this file. -->
-
+{{ defineDatasource "config" .Env.DATASOURCE }}
```
This is what I meant we could do https://github.com/cloudposse/geodesic/blob/master/rootfs/templates/kops/default.yaml#L57

vadim-gleyf committed:
fix(template, Makefile): use inline regexp

osterman commented on modules/readme/Makefile:
```diff
@@ -3,6 +3,9 @@ export README_FILE ?= README.md
 export README_TEMPLATE_FILE ?= $(BUILD_HARNESS_PATH)/templates/README.md
 export README_TEMPLATE_YAML ?= $(BUILD_HARNESS_PATH)/templates/README.yaml
+export USED_DATASOURCE ?= $(if $(datasource),$(datasource),$(shell pwd)/$(README_YAML))
+export DATASOURCE ?= $(if $(datasource),$(datasource),$(README_YAML))
```
where is lowercase `datasource` ever defined?

vadim-gleyf replied:
@osterman for overriding default datasource (`README.yaml` in root folder)
![image](https://user-images.githubusercontent.com/11632146/45751843-e6f4bf00-bc13-11e8-922f-ba67d49371d3.png)

osterman replied:
Is there a way we could just get rid of these envs and define it in the template directly?

vadim-gleyf replied:
@osterman now if no any env passed, behaviour will be as earlier, will be used `README.yaml` from root project folder. 
This env just for possibility to change datasource, it isn't use case?

vadim-gleyf replied:
@osterman move datasource definition from makefile, but gomplate doesn't support inline shell execution, 
`export INCLUDES ?= $(file://$(shell pwd)/?type=text/plain)`
still here

osterman commented on modules/readme/Makefile:
```diff
@@ -29,7 +32,5 @@ readme/lint:
-		--datasource includes='file://$(shell pwd)/?type=text/plain'
-	@echo "Generated $(README_FILE) from $(README_TEMPLATE_FILE) using data from $(README_TEMPLATE_YAML)"
+		--out $(README_FILE) 
+	@echo "Generated $(README_FILE) from $(README_TEMPLATE_FILE) using data from $(USED_DATASOURCE)"
```
What's wrong with using `README_TEMPLATE_YAML`? I don't understand why we need a new variable.

vadim-gleyf replied:
@osterman due possible overriding, it is not necessary?

osterman commented on templates/README.md:
```diff
@@ -1,5 +1,6 @@
 <!-- This file was automatically generated by the `build-harness`. Make all changes to `README.yaml` and run `make readme` to rebuild this file. -->
-
+{{ defineDatasource "config" .Env.DATASOURCE | regexp.Replace ".*" ""}}
```
add a space between the quote and closing brace. `"}}`

osterman commented on templates/README.md:
```diff
@@ -1,5 +1,6 @@
 <!-- This file was automatically generated by the `build-harness`. Make all changes to `README.yaml` and run `make readme` to rebuild this file. -->
-
+{{ defineDatasource "config" .Env.DATASOURCE | regexp.Replace ".*" ""}}
+{{ defineDatasource "includes" .Env.INCLUDES | regexp.Replace ".*" ""}} 
```
add a space between the quote and closing brace. `"}}`

vadim-gleyf committed:
style(README): fix spaces

vadim-gleyf committed:
Merge branch 'master' into dynamic-datasources

osterman reviewed (changes requested):
Can we get rid of the envs for the template and just hard code something in the template? My hope is to make the template almost entirely self-contained.

vadim-gleyf committed:
refactor(readme): mv datasource decl from makefile

osterman committed:
Merge branch 'master' into dynamic-datasources

osterman commented on modules/readme/Makefile:
```diff
@@ -3,6 +3,8 @@ export README_FILE ?= README.md
 export README_YAML ?= README.yaml
 export README_TEMPLATE_FILE ?= $(BUILD_HARNESS_PATH)/templates/README.md
 export README_TEMPLATE_YAML ?= $(BUILD_HARNESS_PATH)/templates/README.yaml
+export USED_DATASOURCE ?= $(if $(datasource),$(datasource),$(shell pwd)/$(README_YAML))
```
Where does is `$(datasource)` defined? I would expect to see `datasource =` somewhere.

osterman replied:
The user can just specify `make readme README_TEMPLATE_YAML=foobar.yaml` if they want to override, so unsure what `datasource` gives us.

osterman reviewed (changes requested):
See comments

vadim-gleyf committed:
refactor(readme): remove useless variable

vadim-gleyf committed:
Merge branch 'dynamic-datasources' of github.com:cloudposse/build-harness into dynamic-datasources

osterman commented on modules/readme/Makefile:
```diff
@@ -2,8 +2,7 @@ export README_LINT ?= $(TMP)/README.md
-export README_TEMPLATE_YAML ?= $(BUILD_HARNESS_PATH)/templates/README.yaml
-export USED_DATASOURCE ?= $(if $(datasource),$(datasource),$(shell pwd)/$(README_YAML))
+export README_TEMPLATE_YAML := $(if $(findstring http,$(README_YAML)),$(README_YAML),$(BUILD_HARNESS_PATH)/templates/$(README_YAML))
 export INCLUDES ?= $(file://$(shell pwd)/?type=text/plain)
```
Rename to `README_INCLUDES` so we don't overload this env.

osterman reviewed (changes requested):
Rename `INCLUDES`

vadim-gleyf committed:
refactor(*): rename INCLUDES var

osterman commented on modules/readme/Makefile:
```diff
@@ -3,7 +3,7 @@ export README_FILE ?= README.md
 export README_TEMPLATE_FILE ?= $(BUILD_HARNESS_PATH)/templates/README.md
 export README_TEMPLATE_YAML := $(if $(findstring http,$(README_YAML)),$(README_YAML),$(BUILD_HARNESS_PATH)/templates/$(README_YAML))
-export INCLUDES ?= $(file://$(shell pwd)/?type=text/plain)
+export FILES ?= $(file://$(shell pwd)/?type=text/plain)
```
No, rename to `README_INCLUDES` - the objective is to "namespace" environment variables related to the readme generation so they don't conflict. Generic names like `INCLUDES` or `FILES` are likely to "overload" other envs.

osterman reviewed (changes requested):
See comments

vadim-gleyf committed:
refactor(README): rename variable

osterman commented on modules/readme/Makefile:
```diff
@@ -2,7 +2,8 @@ export README_LINT ?= $(TMP)/README.md
 export README_YAML ?= README.yaml
 export README_TEMPLATE_FILE ?= $(BUILD_HARNESS_PATH)/templates/README.md
-export README_TEMPLATE_YAML ?= $(BUILD_HARNESS_PATH)/templates/README.yaml
+export README_TEMPLATE_YAML := $(if $(findstring http,$(README_YAML)),$(README_YAML),$(BUILD_HARNESS_PATH)/templates/$(README_YAML))
```
Can you elaborate on why we have findstring here?

vadim-gleyf replied:
@osterman to check is datasource external or not

osterman reviewed (commented):
Explain need for `findstring`
